examples/fabric-admin/device_manager/DeviceManager.cpp

```diff
@@ -0,0 +1,226 @@
+        path.mEndpointId, ChipLogValueMEI(path.mClusterId), ChipLogValueMEI(path.mAttributeId), path.mDataVersion.ValueOr(0));
+
+    app::DataModel::DecodableList<EndpointId> value;
+    CHIP_ERROR error = app::DataModel::Decode(*data, value);
+    if (error != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to decode attribute value. Error: %" CHIP_ERROR_FORMAT, error.Format());
+        return;
+    }
+
+    std::set<EndpointId> newEndpoints;
+
+    // Populate the newEndpoints set from the decoded value using an iterator
+    auto iter = value.begin();
+    while (iter.Next())
+    {
+        newEndpoints.insert(iter.GetValue());
+    }
+
+    if (iter.GetStatus() != CHIP_NO_ERROR)
+    {
+        ChipLogError(NotSpecified, "Failed to iterate over decoded attribute value.");
+        return;
+    }
+
+    // Compare newEndpoints with mSyncedEndpoints to determine added and removed endpoints
+    std::vector<EndpointId> addedEndpoints;
+    std::vector<EndpointId> removedEndpoints;
+
+    // Find added endpoints
+    for (const auto & endpoint : newEndpoints)
+    {
+        if (FindDeviceByEndpoint(endpoint) == nullptr)
+        {
+            addedEndpoints.push_back(endpoint);
+        }
+    }
+
+    // Find removed endpoints
+    for (auto & device : mSyncedDevices)
+    {
+        EndpointId endpointId = device.GetEndpointId();
+        if (newEndpoints.find(endpointId) == newEndpoints.end())
+        {
+            removedEndpoints.push_back(endpointId);
+        }
+    }
+
+    // Process added endpoints
+    for (const auto & endpoint : addedEndpoints)
+    {
+        ChipLogProgress(NotSpecified, "Endpoint added: %u", endpoint);
+
+        if (mAutoSyncEnabled)
+        {
+            char command[64];
+            snprintf(command, sizeof(command), "fabricsync sync-device %d", endpoint);
+
+            usleep(kSyncPrepareTimeMs * 1000);
+            PushCommand(command);
+        }
+    }
+
+    // Process removed endpoints
+    for (const auto & endpoint : removedEndpoints)
+    {
+        ChipLogProgress(NotSpecified, "Endpoint removed: %u", endpoint);
+
+        Device * device = FindDeviceByEndpoint(endpoint);
+
+        if (device == nullptr)
+        {
+            ChipLogProgress(NotSpecified, "No device on Endpoint: %u", endpoint);
+            continue;
+        }
+
+        if (mAutoSyncEnabled)
+        {
+            char command[64];
+            snprintf(command, sizeof(command), "pairing unpair %ld", device->GetNodeId());
+
+            PairingCommand * pairingCommand = static_cast<PairingCommand *>(CommandMgr().GetCommandByName("pairing", "unpair"));
+
+            if (pairingCommand == nullptr)
+            {
+                ChipLogError(NotSpecified, "Pairing code command is not available");
+                return;
+            }
+
+            pairingCommand->RegisterPairingDelegate(this);
+
+            usleep(kSyncPrepareTimeMs * 1000);
```

Comment:
What thread is this sleeping on? If matter thread can we not schedule this?

Reply:
matter thread, lets remove it and see